compiler/ast.nim

```diff
@@ -299,6 +299,7 @@ type
     sfUsedInFinallyOrExcept  # symbol is used inside an 'except' or 'finally'
     sfSingleUsedTemp  # For temporaries that we know will only be used once
     sfNoalias         # 'noalias' annotation, means C's 'restrict'
+    sfEffectsDelayed  # an 'effectsDelayed' parameter
```

Comment:
the bugs i mentioned in https://github.com/nim-lang/Nim/pull/18777#issuecomment-910774236 is still there, it looks like the compiler ignores the `effectsOf: p1` annotations and assumes every param is called

Reply:
Your snippet was even added as a test case, it works...

Reply:
ok, i was missing `{.experimental: "strictEffects".}`; however, the problem remains if i remove the explicit annotaton:

```nim
when defined case9:
  {.push warningAsError[Effect]: on.}
  {.experimental: "strictEffects".}
  # proc fn(a: int, p1, p2: proc()) {.effectsOf: p1.} =
  # proc fn(a: int, p1, p2: proc()) {.effectsOf: p2.} =
  proc fn(a: int, p1, p2: proc()) =
    if a == 7:
      p1()
    if a<0:
      raise newException(ValueError, $a)

  proc main() {.raises: [ValueError].} =
    fn(1, proc()=discard, proc() = raise newException(IOError, "foo"))
  main()

```

this gives:

t12754.nim(111, 7) Error: fn(1, proc () = discard , proc () = raise
  (ref IOError)(msg: "foo", parent: nil)) can raise an unlisted exception: Exception [Effect]


but compiler should infer that this is `{.effectsOf: p1.}` since only `p1` is called, not `p2`, and then it should not issue any warning since main can't raise anything other than ValueError

Reply:
There is no inference for the `.effectsOf` pragma and there is no plan to add any, it would repeat the mistakes of "sink inference".